open_feature/provider/in_memory_provider.py

```diff
@@ -0,0 +1,120 @@
+from dataclasses import dataclass
+import typing
+
+from open_feature.evaluation_context.evaluation_context import EvaluationContext
+from open_feature.exception.error_code import ErrorCode
+from open_feature.flag_evaluation.reason import Reason
+from open_feature.flag_evaluation.resolution_details import FlagResolutionDetails
+from open_feature.hooks.hook import Hook
+from open_feature.provider.metadata import Metadata
+from open_feature.provider.provider import AbstractProvider
+
+PASSED_IN_DEFAULT = "Passed in default"
+
+
+@dataclass
+class InMemoryMetadata(Metadata):
+    name: str = "In-Memory Provider"
+
+
+T = typing.TypeVar("T", covariant=True)
+
+
+@dataclass
+class InMemoryFlag(typing.Generic[T]):
+    flag_key: str
+    default_variant: str
+    variants: dict[str, T]
+    reason: typing.Optional[Reason] = Reason.STATIC
+    error_code: typing.Optional[ErrorCode] = None
+    error_message: typing.Optional[str] = None
+    context_evaluator: typing.Optional[
+        typing.Callable[["InMemoryFlag", EvaluationContext], FlagResolutionDetails[T]]
+    ] = None
+
+    def resolve(
+        self, evaluation_context: typing.Optional[EvaluationContext]
+    ) -> FlagResolutionDetails[T]:
+        if evaluation_context and self.context_evaluator:
+            return self.context_evaluator(self, evaluation_context)
+
+        return FlagResolutionDetails(
+            value=self.variants[self.default_variant],
+            reason=self.reason,
+            variant=self.default_variant,
+            error_code=self.error_code,
+            error_message=self.error_message,
+        )
+
+
+FlagStorage = typing.Dict[str, InMemoryFlag]
+
+V = typing.TypeVar("V")
+
+
+class InMemoryProvider(AbstractProvider):
+    _flags: FlagStorage
+
+    def __init__(self, flags: FlagStorage):
```

Comment:
My worry with this API is that it's very tempting for users to retain a copy of the mutable data structure behind our backs, which could cause difficult to debug bugs.

Might be a bigger problems in languages without a GIL.

Reply:
could we clone the object so that such mutation isn't possible?